misc-utils/lsclocks.c

```diff
@@ -0,0 +1,366 @@
+ *
+ * This program is distributed in the hope that it would be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include <stdio.h>
+#include <stdbool.h>
+#include <time.h>
+#include <inttypes.h>
+#include <getopt.h>
+
+#include <libsmartcols.h>
+
+#include "c.h"
+#include "nls.h"
+#include "strutils.h"
+#include "timeutils.h"
+#include "closestream.h"
+#include "xalloc.h"
+
+#define CLOCKFD 3
+#define FD_TO_CLOCKID(fd) ((~(clockid_t) (fd) << 3) | CLOCKFD)
+
+#ifndef CLOCK_REALTIME
+#define CLOCK_REALTIME			0
+#endif
+
+#ifndef CLOCK_MONOTONIC
+#define CLOCK_MONOTONIC			1
+#endif
+
+#ifndef CLOCK_MONOTONIC_RAW
+#define CLOCK_MONOTONIC_RAW		4
+#endif
+
+#ifndef CLOCK_REALTIME_COARSE
+#define CLOCK_REALTIME_COARSE		5
+#endif
+
+#ifndef CLOCK_MONOTONIC_COARSE
+#define CLOCK_MONOTONIC_COARSE		6
+#endif
+
+#ifndef CLOCK_BOOTTIME
+#define CLOCK_BOOTTIME			7
+#endif
+
+#ifndef CLOCK_REALTIME_ALARM
+#define CLOCK_REALTIME_ALARM		8
+#endif
+
+#ifndef CLOCK_BOOTTIME_ALARM
+#define CLOCK_BOOTTIME_ALARM		9
+#endif
+
+#ifndef CLOCK_TAI
+#define CLOCK_TAI			11
+#endif
+
+struct clockinfo {
+	clockid_t id;
+	const char * const id_name;
+	const char * const name;
+};
+
+static const struct clockinfo clocks[] = {
+	{ CLOCK_REALTIME,         "CLOCK_REALTIME",         "realtime"         },
+	{ CLOCK_MONOTONIC,        "CLOCK_MONOTONIC",        "monotonic"        },
+	{ CLOCK_MONOTONIC_RAW,    "CLOCK_MONOTONIC_RAW",    "monotonic-raw"    },
+	{ CLOCK_REALTIME_COARSE,  "CLOCK_REALTIME_COARSE",  "realtime-coarse"  },
+	{ CLOCK_MONOTONIC_COARSE, "CLOCK_MONOTONIC_COARSE", "monotonic-coarse" },
+	{ CLOCK_BOOTTIME,         "CLOCK_BOOTTIME",         "boottime"         },
+	{ CLOCK_REALTIME_ALARM,   "CLOCK_REALTIME_ALARM",   "realtime-alarm"   },
+	{ CLOCK_BOOTTIME_ALARM,   "CLOCK_BOOTTIME_ALARM",   "boottime-alarm"   },
+	{ CLOCK_TAI,              "CLOCK_TAI",              "tai"              },
+};
+
+/* column IDs */
+enum {
+	COL_ID,
+	COL_NAME,
+	COL_HUMAN_NAME,
+	COL_TIME,
+	COL_ISO_TIME,
+	COL_RESOLUTION,
+};
+
+/* column names */
+struct colinfo {
+	const char * const	name; /* header */
+	double			whint; /* width hint (N < 1 is in percent of termwidth) */
+	int			flags; /* SCOLS_FL_* */
+	int			json_type; /* SCOLS_JSON_* */
+	const char * const	help;
+};
+
+/* columns descriptions */
+static const struct colinfo infos[] = {
+	[COL_ID]         = { "ID",         1, SCOLS_FL_RIGHT, SCOLS_JSON_NUMBER, N_("numeric id") },
+	[COL_NAME]       = { "NAME",       1, 0,              SCOLS_JSON_STRING, N_("symbolic id") },
+	[COL_HUMAN_NAME] = { "HUMAN_NAME", 1, 0,              SCOLS_JSON_STRING, N_("human name") },
```

Comment:
I'm not sure about the column names NAME and HUMAN_NAME. What about IDNAME (or TYPE) and NAME?